From a6168a2f3fa5aa86053adefcb64fa1d95abe3284 Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Thu, 28 Aug 2025 16:48:37 +0200
Subject: [PATCH 003/553] CMake: Workaround Android CMP0155 issue

If a user:
- targets Android
- sets cmake_minimum_required(3.29) in their project which implies
  policy CMP0155 being set to NEW
- sets CMAKE_CXX_STANDARD >= 20
- uses Qt Creator's maintenance tool provider
- uses Qt Creator's default of ANDROID_USE_LEGACY_TOOLCHAIN_FILE=OFF

then they will hit an issue where the Threads package cannot be found
due to a CMake bug that causes all try_compile calls to fail, which
in turns causes the Qt6 package not being found.

Android and ANDROID_USE_LEGACY_TOOLCHAIN_FILE being OFF is relevant,
because in that case CMake fails to find the
CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS tool, which is used during
try_compile builds.

CMP0155 set to NEW and CMAKE_CXX_STANDARD >= 20 is relevant,
because only in that case will CMake try to use the scan deps tool to
scan for modules during try_compile builds.

The maintenance tool provider is relevant in this case, because CMake
records the CMP0155 policy value at function definition time.

Normally without a dependency provider, the policy value at Threads
package lookup time would be OLD because each Qt6Config.cmake file
has a cmake_minimum_required(3.16..3.21) call which overrides any
other policy value.
But because the dependency provider function is defined before any of
Qt's code is loaded, it inherits the CMP0155 NEW from
cmake_minimum_required(3.29) of the user's project. That causes
all find_package lookups to use CMP0155=NEW, ignoring Qt's request.
This works by design from CMake's PoV, but causes unfortunate fallout
in this case.

Because we can't influence the recorded policy value of the provided
function from the outside in any way, the only thing we can do is
disable module scanning via CMAKE_CXX_SCAN_FOR_MODULES.

Work around the issue by setting CMAKE_CXX_SCAN_FOR_MODULES to OFF
inside Qt6Config.cmake, which will set it in the directory scope of
the find_package(Qt6) caller, affecting all try_compile calls there.
Do it only in the specific conditions mentioned above.

Allow opting out of the variable assignment by configuring with
-DQT_NO_SET_CMAKE_CXX_SCAN_FOR_MODULES_TO_OFF=ON

Doing this in Qt6Config.cmake allows working around the issue when
using older Creator versions that don't work around it themselves,

See https://gitlab.kitware.com/cmake/cmake/-/issues/27169

Pick-to: 6.9 6.8
Fixes: QTBUG-139439
Change-Id: Id171c88566c41b65b24c2f2a3874e3f3c0abc814
Reviewed-by: Alexey Edelev <alexey.edelev@qt.io>
(cherry picked from commit f257009870f0016d6c48b23e3f41228f4b8158c3)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtBaseGlobalTargets.cmake    |  8 ++++++++
 cmake/QtConfig.cmake.in            |  2 ++
 cmake/QtPublicAndroidHelpers.cmake | 17 +++++++++++++++++
 3 files changed, 27 insertions(+)

diff --git a/cmake/QtBaseGlobalTargets.cmake b/cmake/QtBaseGlobalTargets.cmake
index 9444bb4644e..041e6a0d7b2 100644
--- a/cmake/QtBaseGlobalTargets.cmake
+++ b/cmake/QtBaseGlobalTargets.cmake
@@ -204,6 +204,14 @@ qt_internal_get_max_new_policy_cmake_version(max_new_policy_version)
 qt_internal_get_qt_build_public_helpers(__qt_cmake_public_helpers)
 list(JOIN __qt_cmake_public_helpers "\n    " QT_PUBLIC_FILES_TO_INCLUDE)
 
+set(__qt_cmake_extra_code_before_dependencies "")
+if(ANDROID)
+    list(APPEND __qt_cmake_extra_code_before_dependencies
+        "__qt_internal_workaround_android_cmp0155_issue()")
+endif()
+list(JOIN __qt_cmake_extra_code_before_dependencies
+    "\n    " QT_CONFIG_EXTRA_CODE_BEFORE_DEPENDENCIES)
+
 # Generate and install Qt6 config file. Make sure it happens after the global feature evaluation so
 # they can be accessed in the Config file if needed.
 configure_package_config_file(
diff --git a/cmake/QtConfig.cmake.in b/cmake/QtConfig.cmake.in
index 2c34af97713..c1d52375f77 100644
--- a/cmake/QtConfig.cmake.in
+++ b/cmake/QtConfig.cmake.in
@@ -113,6 +113,8 @@ endif()
 # Mark that the current directory scope has its sanitizer flags set.
 set(__qt_sanitizer_options_set TRUE)
 
+@QT_CONFIG_EXTRA_CODE_BEFORE_DEPENDENCIES@
+
 # Find required dependencies, if any.
 include(CMakeFindDependencyMacro)
 if(EXISTS "${CMAKE_CURRENT_LIST_DIR}/@INSTALL_CMAKE_NAMESPACE@Dependencies.cmake")
diff --git a/cmake/QtPublicAndroidHelpers.cmake b/cmake/QtPublicAndroidHelpers.cmake
index 4baba9f88aa..30ffec43ad8 100644
--- a/cmake/QtPublicAndroidHelpers.cmake
+++ b/cmake/QtPublicAndroidHelpers.cmake
@@ -1,6 +1,23 @@
 # Copyright (C) 2024 The Qt Company Ltd.
 # SPDX-License-Identifier: BSD-3-Clause
 
+function(__qt_internal_workaround_android_cmp0155_issue)
+    # Work around upstream cmake issue: https://gitlab.kitware.com/cmake/cmake/-/issues/27169
+    if(ANDROID
+        AND CMAKE_VERSION VERSION_GREATER_EQUAL 3.29
+        AND NOT ANDROID_USE_LEGACY_TOOLCHAIN_FILE
+        AND NOT CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS
+        AND NOT QT_NO_SET_CMAKE_CXX_SCAN_FOR_MODULES_TO_OFF
+      )
+      message(DEBUG
+        "Setting CMAKE_CXX_SCAN_FOR_MODULES to OFF in the Qt6 package directory scope to "
+        "avoid issues with not being able to find the Threads package when targeting Android with "
+        "cmake_minimum_required(3.29) and CMAKE_CXX_STANDARD >= 20."
+      )
+      set(CMAKE_CXX_SCAN_FOR_MODULES OFF PARENT_SCOPE)
+    endif()
+endfunction()
+
 function(_qt_internal_detect_latest_android_platform out_var)
     # Locate the highest available platform
     file(GLOB android_platforms
-- 
2.51.2

